zcash_proofs/src/sapling/verifier/batch.rs

```diff
@@ -0,0 +1,175 @@
+                },
+                |this, proof, public_inputs| {
+                    this.spend_proofs.queue((proof, public_inputs.to_vec()));
+                    true
+                },
+            );
+            if !consensus_rules_passed {
+                return false;
+            }
+        }
+
+        for output in bundle.shielded_outputs {
+            // Deserialize the ephemeral key
+            let epk = match jubjub::ExtendedPoint::from_bytes(&output.ephemeral_key.0).into() {
+                Some(p) => p,
+                None => return false,
+            };
+
+            // Deserialize the proof
+            let zkproof = match groth16::Proof::read(&output.zkproof[..]) {
+                Ok(p) => p,
+                Err(_) => return false,
+            };
+
+            // Check the Output consensus rules, and batch its proof.
+            let consensus_rules_passed = ctx.check_output(
+                output.cv,
+                output.cmu,
+                epk,
+                zkproof,
+                |proof, public_inputs| {
+                    self.output_proofs.queue((proof, public_inputs.to_vec()));
+                    true
+                },
+            );
+            if !consensus_rules_passed {
+                return false;
+            }
+        }
+
+        // Check the whole-bundle consensus rules, and batch the binding signature.
+        ctx.final_check(
+            bundle.value_balance,
+            &sighash,
+            bundle.authorization.binding_sig,
+            |bvk, _, binding_sig| {
+                let bvk =
+                    redjubjub::VerificationKeyBytes::<redjubjub::Binding>::from(bvk.0.to_bytes());
+                let binding_sig = {
+                    let mut buf = [0; 64];
+                    binding_sig.write(&mut buf[..]).unwrap();
+                    redjubjub::Signature::<redjubjub::Binding>::from(buf)
+                };
+
+                self.signatures.queue((bvk, binding_sig, &sighash));
+                true
+            },
+        )
+    }
+
+    /// Batch-validates the accumulated bundles.
+    ///
+    /// Returns `true` if every proof and signature in every bundle added to the batch
+    /// validator is valid, or `false` if one or more are invalid. No attempt is made to
+    /// figure out which of the accumulated bundles might be invalid; if that information
+    /// is desired, construct separate [`BatchValidator`]s for sub-batches of the bundles.
+    pub fn validate<R: RngCore + CryptoRng>(
+        self,
+        spend_vk: &groth16::VerifyingKey<Bls12>,
+        output_vk: &groth16::VerifyingKey<Bls12>,
+        mut rng: R,
+    ) -> bool {
+        if !self.bundles_added {
+            // An empty batch is always valid, but is not free to run; skip it. Note that
+            // a bundle has at least a binding signature, so if there are no binding
+            // signatures, there are also no proofs or spend authorization signatures.
```

Comment:
```suggestion
            // An empty batch is always valid, but is not free to run; skip it.
```
We're not relying on a bundle having at least a binding signature here.